File: reference-kits/semiconductor-predictive-maintenance/demo/images/edge-inference/app/health_model.py
"""
health_model.py — on-device predictive-maintenance model for a semiconductor
plasma-etch chamber.

This is a *real* streaming anomaly-detection model, not a mock. It runs on a CPU
in a few kilobytes of state per tool, which is why it fits a single-node edge box.
The production hand-off swaps this for a heavier model served by SUSE AI (Ollama /
vLLM + Milvus on a GPU node) — the interface (sensor frame in, health verdict out)
stays identical, which is the whole point of the demo.

Technique — an EWMA-smoothed multivariate control chart (textbook Hotelling T² /
Mahalanobis SPC, the standard fab statistical-process-control approach):

  1. Baseline learning. For the first `warmup` frames a tool is assumed healthy
     (standard commissioning assumption). We accumulate an exact per-sensor mean
     and variance with Welford's algorithm, then FREEZE that baseline. No training
     data set ships with the model — each tool learns its own baseline in situ.
  2. Per-sensor z-score against the frozen baseline, clipped to +/-Z_CLIP so one
     saturated channel can't dominate.
  3. Mahalanobis-style statistic m = sum(z_i^2) over the sensors present. Under a
     healthy tool E[m] = k (the sensor count); a fault inflates it.
  4. EWMA-smooth m (fast for the live health score, slow for the trend) so healthy
     measurement noise does not trip an alarm but a sustained shift does.
  5. Health 0-100 = 100 * exp(-anomaly / DECAY), where anomaly = sqrt(excess/k)
     and excess = max(0, m_smoothed - k). Healthy => excess ~ 0 => health ~ 100.
  6. Remaining-useful-life (RUL): the slow EWMA gives a degradation trend; RUL =
     frames until projected health crosses CRITICAL, by linear extrapolation. A
     flat / improving trend reports RUL = None (nominal, beyond horizon).

Everything here is deterministic and unit-tested (see demo/tests/).
"""
import math
from collections import OrderedDict, deque

# --- Sensor contract (must match the simulator; see demo/README.md) -----------
SENSORS = (
    "chamber_pressure_mtorr",
    "rf_forward_power_w",
    "rf_reflected_power_w",
    "chamber_temp_c",
    "endpoint_signal",
    "he_backside_flow_sccm",
)

# --- Verdict states (by health score 0-100) -----------------------------------
STATE_HEALTHY = "HEALTHY"
STATE_WATCH = "WATCH"
STATE_WARNING = "WARNING"
STATE_CRITICAL = "CRITICAL"

_STATE_BANDS = (
    (85.0, STATE_HEALTHY),
    (65.0, STATE_WATCH),
    (40.0, STATE_WARNING),
)
CRITICAL_HEALTH = 40.0

# --- Model constants ----------------------------------------------------------
DECAY = 2.2          # health = 100*exp(-anomaly/DECAY)
Z_CLIP = 8.0         # clip per-sensor z-scores
FAST_ALPHA = 0.15    # EWMA weight for the live health statistic
SLOW_ALPHA = 0.03    # EWMA weight for the degradation-trend statistic
UCL_SIGMA = 2.0      # control-limit offset (sigmas of the statistic's noise)
STD_FLOOR = 1e-3     # guard against a zero-variance (constant) sensor
EPS = 1e-9
DEFAULT_WARMUP = 30
SLOPE_WINDOW = 20    # frames of trend history used to estimate the RUL slope
FORECAST_HORIZON = 100000  # frames; a RUL above this is reported as nominal


def classify(health):
    """Map a 0-100 health score to a discrete state."""
    for threshold, state in _STATE_BANDS:
        if health >= threshold:
            return state
    return STATE_CRITICAL


class _Welford:
    """Exact online mean/variance for one sensor during the warmup window."""

    __slots__ = ("n", "mean", "m2", "std")

    def __init__(self):
        self.n = 0
        self.mean = 0.0
        self.m2 = 0.0
        self.std = STD_FLOOR

    def observe(self, x):
        self.n += 1
        delta = x - self.mean
        self.mean += delta / self.n
        self.m2 += delta * (x - self.mean)

    def freeze(self):
        var = self.m2 / (self.n - 1) if self.n > 1 else 0.0
        self.std = max(math.sqrt(var), STD_FLOOR)

    def zscore(self, x):
        return (x - self.mean) / self.std


class ToolHealthModel:
    """One instance per physical tool. Feed sensor frames, read verdicts."""

    def __init__(self, tool_id, warmup=DEFAULT_WARMUP):
        self.tool_id = tool_id
        self.warmup = warmup
        self.frames = 0
        self._base = OrderedDict((s, _Welford()) for s in SENSORS)
        self._frozen = False
        self._k = len(SENSORS)          # active sensor count, set at freeze
        self._m_fast = float(len(SENSORS))
        self._m_slow = float(len(SENSORS))
        self._slow_hist = deque(maxlen=SLOPE_WINDOW)

    def _mahal(self, frame):
        acc = 0.0
        k = 0
        contribs = {}
        for name, base in self._base.items():
            if name not in frame:
                continue
            z = base.zscore(float(frame[name]))
            z = max(-Z_CLIP, min(Z_CLIP, z))
            contribs[name] = z
            acc += z * z
            k += 1
        return acc, k, contribs

    def update(self, frame):
        """Ingest one sensor frame (dict name->value); return a verdict dict."""
        self.frames += 1

        # --- warmup: learn & freeze baseline, report healthy -----------------
        if self.frames <= self.warmup:
            for name, base in self._base.items():
                if name in frame:
                    base.observe(float(frame[name]))
            if self.frames == self.warmup:
                active = 0
                for base in self._base.values():
                    base.freeze()
                    if base.n > 0:
                        active += 1
                self._k = max(active, 1)
                self._m_fast = float(self._k)
                self._m_slow = float(self._k)
                self._slow_hist.clear()
                self._frozen = True
            return self._verdict(100.0, 0.0, None, [], warming=True)

        # --- scoring against frozen baseline ---------------------------------
        mahal, k, contribs = self._mahal(frame)
        k = k or self._k
        self._m_fast = (1 - FAST_ALPHA) * self._m_fast + FAST_ALPHA * mahal
        self._m_slow = (1 - SLOW_ALPHA) * self._m_slow + SLOW_ALPHA * mahal
        self._slow_hist.append(self._m_slow)

        anomaly = self._anomaly(self._m_fast, k)
        health = max(0.0, min(100.0, 100.0 * math.exp(-anomaly / DECAY)))
        rul = self._rul(k)

        top = sorted(contribs.items(), key=lambda kv: -abs(kv[1]))[:3]
        top_list = [{"sensor": n, "z": round(z, 2)} for n, z in top]
        return self._verdict(health, anomaly, rul, top_list, warming=False)

    @staticmethod
    def _anomaly(m_smoothed, k):
        # Under a healthy tool the statistic fluctuates around k with a known
        # noise floor (the EWMA variance of a chi-square(k) input). Subtract that
        # floor so healthy noise scores ~0 and only a genuine shift registers —
        # this is the upper-control-limit offset of a Hotelling T-squared chart.
        if not k:
            return 0.0
        noise_std = math.sqrt(FAST_ALPHA / (2.0 - FAST_ALPHA) * 2.0 * k)
        excess = max(0.0, m_smoothed - k - UCL_SIGMA * noise_std)
        return math.sqrt(excess / k)

    def _rul(self, k):
        # Estimate the degradation slope by least-squares over the trend window,
        # in anomaly-space (the quantity health is a function of). A single-step
        # difference of a slow EWMA is too noisy to forecast from.
        hist = self._slow_hist
        if len(hist) < SLOPE_WINDOW:
            return None
        ys = [self._anomaly(m, k) for m in hist]
        n = len(ys)
        xbar = (n - 1) / 2.0
        ybar = sum(ys) / n
        num = sum((i - xbar) * (ys[i] - ybar) for i in range(n))
        den = sum((i - xbar) ** 2 for i in range(n))
        slope = num / den if den > EPS else 0.0
        if slope <= EPS:                     # flat or improving -> nominal
            return None
        cur = ys[-1]
        anom_crit = -DECAY * math.log(CRITICAL_HEALTH / 100.0)
        if cur >= anom_crit:
            return 0
        frames = (anom_crit - cur) / slope
        if frames >= FORECAST_HORIZON or frames < 0:
            return None
        return int(round(frames))

    def _verdict(self, health, anomaly, rul, top_list, warming):
        health = round(health, 2)
        return {
            "tool_id": self.tool_id,
            "frames": self.frames,
            "warming": warming,
            "health": health,
            "state": STATE_HEALTHY if warming else classify(health),
            "anomaly": round(anomaly, 4),
            "rul_frames": rul,
            "top_contributors": top_list,
        }


class Fleet:
    """Holds many tools; used by the inference service and the dashboard."""

    def __init__(self, warmup=DEFAULT_WARMUP):
        self._warmup = warmup
        self._models = {}
        self.last_verdict = {}

    def ingest(self, tool_id, frame):
        m = self._models.get(tool_id)
        if m is None:
            m = ToolHealthModel(tool_id, warmup=self._warmup)
            self._models[tool_id] = m
        verdict = m.update(frame)
        self.last_verdict[tool_id] = verdict
        return verdict

    def snapshot(self):
        """Sorted worst-first — what the dashboard renders."""
        order = {STATE_CRITICAL: 0, STATE_WARNING: 1, STATE_WATCH: 2,
                 STATE_HEALTHY: 3}
        return sorted(
            self.last_verdict.values(),
            key=lambda v: (order.get(v["state"], 9), v["health"]),
        )

File: portal/lib/demo/health-model.ts
/**
 * health-model.ts — TypeScript port of the on-device predictive-maintenance
 * model for a semiconductor plasma-etch chamber.
 *
 * This is a faithful, line-for-line port of the Python original that runs in
 * the k3s reference kit:
 *   reference-kits/semiconductor-predictive-maintenance/demo/images/
 *     edge-inference/app/health_model.py
 * Same technique — an EWMA-smoothed multivariate control chart (Hotelling T² /
 * Mahalanobis SPC), the standard fab statistical-process-control approach:
 *
 *   1. Baseline learning: per-sensor mean/variance via Welford's algorithm for
 *      the first `warmup` frames, then FROZEN. Each tool learns its own
 *      baseline in situ; no training set ships with the model.
 *   2. Per-sensor z-score against the frozen baseline, clipped to ±Z_CLIP.
 *   3. Mahalanobis-style statistic m = Σ z² over the sensors present.
 *   4. EWMA smoothing (fast → live health, slow → degradation trend).
 *   5. health = 100·exp(−anomaly/DECAY), anomaly = √(excess/k) above the
 *      2σ upper control limit.
 *   6. RUL = frames until projected health crosses CRITICAL, by least-squares
 *      extrapolation over the slow-EWMA trend window.
 *
 * Parity with the Python model is *proven*, not asserted: see
 * tests/demo/golden-parity.test.ts, which replays 590 recorded frames from the
 * real Python model (scripts/generate-golden-vectors.py) through this port and
 * compares every verdict. If you change a constant here, change the Python
 * model too and regenerate the vectors — they are one model in two languages.
 */
import type { Contributor, SensorFrame, ToolState, Verdict } from "./types";

// --- Sensor contract (must match the simulator) ------------------------------
export const SENSORS = [
  "chamber_pressure_mtorr",
  "rf_forward_power_w",
  "rf_reflected_power_w",
  "chamber_temp_c",
  "endpoint_signal",
  "he_backside_flow_sccm",
] as const;

// --- Verdict states (by health score 0-100) ----------------------------------
const STATE_BANDS: Array<[number, ToolState]> = [
  [85.0, "HEALTHY"],
  [65.0, "WATCH"],
  [40.0, "WARNING"],
];
export const CRITICAL_HEALTH = 40.0;

// --- Model constants (identical to the Python model) -------------------------
const DECAY = 2.2; // health = 100·exp(−anomaly/DECAY)
const Z_CLIP = 8.0; // clip per-sensor z-scores
const FAST_ALPHA = 0.15; // EWMA weight, live health statistic
const SLOW_ALPHA = 0.03; // EWMA weight, degradation-trend statistic
const UCL_SIGMA = 2.0; // control-limit offset (σ of the statistic's noise)
const STD_FLOOR = 1e-3; // guard against a zero-variance (constant) sensor
const EPS = 1e-9;
export const DEFAULT_WARMUP = 30;
const SLOPE_WINDOW = 20; // frames of trend history for the RUL slope
const FORECAST_HORIZON = 100000; // frames; beyond this RUL reports nominal

/** Python-style round(x, ndigits): round-half-to-even on the scaled value.
 * (JS Math.round is half-up; Python uses banker's rounding — the golden
 * vectors were produced by Python, so we match its convention.) */
function pyRound(x: number, ndigits: number): number {
  const m = Math.pow(10, ndigits);
  const y = x * m;
  const floor = Math.floor(y);
  const diff = y - floor;
  let r: number;
  if (diff > 0.5) r = floor + 1;
  else if (diff < 0.5) r = floor;
  else r = floor % 2 === 0 ? floor : floor + 1;
  return r / m;
}

/** Map a 0-100 health score to a discrete state. */
export function classify(health: number): ToolState {
  for (const [threshold, state] of STATE_BANDS) {
    if (health >= threshold) return state;
  }
  return "CRITICAL";
}

/** Exact online mean/variance for one sensor during the warmup window. */
class Welford {
  n = 0;
  mean = 0.0;
  m2 = 0.0;
  std = STD_FLOOR;

  observe(x: number): void {
    this.n += 1;
    const delta = x - this.mean;
    this.mean += delta / this.n;
    this.m2 += delta * (x - this.mean);
  }

  freeze(): void {
    const variance = this.n > 1 ? this.m2 / (this.n - 1) : 0.0;
    this.std = Math.max(Math.sqrt(variance), STD_FLOOR);
  }

  zscore(x: number): number {
    return (x - this.mean) / this.std;
  }
}

/** One instance per physical tool. Feed sensor frames, read verdicts. */
export class ToolHealthModel {
  readonly toolId: string;
  readonly warmup: number;
  frames = 0;
  private base = new Map<string, Welford>();
  private k: number = SENSORS.length; // active sensor count, set at freeze
  private mFast: number = SENSORS.length;
  private mSlow: number = SENSORS.length;
  private slowHist: number[] = []; // bounded to SLOPE_WINDOW

  constructor(toolId: string, warmup: number = DEFAULT_WARMUP) {
    this.toolId = toolId;
    this.warmup = warmup;
    for (const s of SENSORS) this.base.set(s, new Welford());
  }

  private mahal(frame: SensorFrame): {
    acc: number;
    k: number;
    contribs: Map<string, number>;
  } {
    let acc = 0.0;
    let k = 0;
    const contribs = new Map<string, number>();
    for (const [name, base] of this.base) {
      if (!(name in frame)) continue;
      let z = base.zscore(frame[name]);
      z = Math.max(-Z_CLIP, Math.min(Z_CLIP, z));
      contribs.set(name, z);
      acc += z * z;
      k += 1;
    }
    return { acc, k, contribs };
  }

  /** Ingest one sensor frame; return a verdict. */
  update(frame: SensorFrame): Verdict {
    this.frames += 1;

    // --- warmup: learn & freeze baseline, report healthy --------------------
    if (this.frames <= this.warmup) {
      for (const [name, base] of this.base) {
        if (name in frame) base.observe(frame[name]);
      }
      if (this.frames === this.warmup) {
        let active = 0;
        for (const base of this.base.values()) {
          base.freeze();
          if (base.n > 0) active += 1;
        }
        this.k = Math.max(active, 1);
        this.mFast = this.k;
        this.mSlow = this.k;
        this.slowHist = [];
      }
      return this.verdict(100.0, 0.0, null, [], true);
    }

    // --- scoring against frozen baseline ------------------------------------
    const { acc: mahal, k: kPresent, contribs } = this.mahal(frame);
    const k = kPresent || this.k;
    this.mFast = (1 - FAST_ALPHA) * this.mFast + FAST_ALPHA * mahal;
    this.mSlow = (1 - SLOW_ALPHA) * this.mSlow + SLOW_ALPHA * mahal;
    this.slowHist.push(this.mSlow);
    if (this.slowHist.length > SLOPE_WINDOW) this.slowHist.shift();

    const anomaly = ToolHealthModel.anomaly(this.mFast, k);
    const health = Math.max(0.0, Math.min(100.0, 100.0 * Math.exp(-anomaly / DECAY)));
    const rul = this.rul(k);

    const top: Contributor[] = [...contribs.entries()]
      .sort((a, b) => Math.abs(b[1]) - Math.abs(a[1]))
      .slice(0, 3)
      .map(([sensor, z]) => ({ sensor, z: pyRound(z, 2) }));
    return this.verdict(health, anomaly, rul, top, false);
  }

  /** Anomaly above the upper control limit: under a healthy tool the statistic
   * fluctuates around k with a known EWMA-of-χ²(k) noise floor; subtract that
   * floor so healthy noise scores ~0 and only a genuine shift registers. */
  private static anomaly(mSmoothed: number, k: number): number {
    if (!k) return 0.0;
    const noiseStd = Math.sqrt((FAST_ALPHA / (2.0 - FAST_ALPHA)) * 2.0 * k);
    const excess = Math.max(0.0, mSmoothed - k - UCL_SIGMA * noiseStd);
    return Math.sqrt(excess / k);
  }

  /** Least-squares degradation slope over the trend window, in anomaly-space;
   * a flat / improving trend reports null (nominal, beyond horizon). */
  private rul(k: number): number | null {
    const hist = this.slowHist;
    if (hist.length < SLOPE_WINDOW) return null;
    const ys = hist.map((m) => ToolHealthModel.anomaly(m, k));
    const n = ys.length;
    const xbar = (n - 1) / 2.0;
    const ybar = ys.reduce((a, b) => a + b, 0) / n;
    let num = 0.0;
    let den = 0.0;
    for (let i = 0; i < n; i++) {
      num += (i - xbar) * (ys[i] - ybar);
      den += (i - xbar) * (i - xbar);
    }
    const slope = den > EPS ? num / den : 0.0;
    if (slope <= EPS) return null;
    const cur = ys[n - 1];
    const anomCrit = -DECAY * Math.log(CRITICAL_HEALTH / 100.0);
    if (cur >= anomCrit) return 0;
    const frames = (anomCrit - cur) / slope;
    if (frames >= FORECAST_HORIZON || frames < 0) return null;
    return Math.round(frames);
  }

  private verdict(
    health: number,
    anomaly: number,
    rul: number | null,
    top: Contributor[],
    warming: boolean,
  ): Verdict {
    const h = pyRound(health, 2);
    return {
      tool_id: this.toolId,
      frames: this.frames,
      warming,
      health: h,
      state: warming ? "HEALTHY" : classify(h),
      anomaly: pyRound(anomaly, 4),
      rul_frames: rul,
      top_contributors: top,
    };
  }
}

/** Holds many tools; the browser engine and dashboard read from this. */
export class Fleet {
  private warmup: number;
  private models = new Map<string, ToolHealthModel>();
  readonly lastVerdict = new Map<string, Verdict>();

  constructor(warmup: number = DEFAULT_WARMUP) {
    this.warmup = warmup;
  }

  ingest(toolId: string, frame: SensorFrame): Verdict {
    let m = this.models.get(toolId);
    if (!m) {
      m = new ToolHealthModel(toolId, this.warmup);
      this.models.set(toolId, m);
    }
    const verdict = m.update(frame);
    this.lastVerdict.set(toolId, verdict);
    return verdict;
  }

  /** Sorted worst-first — what the dashboard renders. */
  snapshot(): Verdict[] {
    const order: Record<ToolState, number> = {
      CRITICAL: 0,
      WARNING: 1,
      WATCH: 2,
      HEALTHY: 3,
    };
    return [...this.lastVerdict.values()].sort(
      (a, b) => order[a.state] - order[b.state] || a.health - b.health,
    );
  }
}
